url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/io/Generic_XMLDOMReader.java

```diff
@@ -0,0 +1,140 @@
+
+/**
+ * Adapted from:
+ * http://www.mkyong.com/java/how-to-read-xml-file-in-java-dom-parser/
+ */
+public abstract class Generic_XMLDOMReader {
+
+    protected File file;
+    protected NodeList nodeList;
+    protected String nodeName;
+    protected DocumentBuilderFactory aDocumentBuilderFactory;
+    protected DocumentBuilder aDocumentBuilder;
+    protected Document aDocument;
+
+    protected void init(
+            File file,
+            String nodeName) {
+        this.file = file;
+        this.nodeName = nodeName;
+        initDocumentBuilderFactory();
+        initDocumentBuilder();
+        initDocument();
+        initNodeList(nodeName);
+    }
+
+    protected void initDocumentBuilderFactory() {
+        aDocumentBuilderFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    protected void initDocumentBuilder() {
+        try {
+            aDocumentBuilder = aDocumentBuilderFactory.newDocumentBuilder();
+        } catch (ParserConfigurationException ex) {
+            Logger.getLogger(Generic_XMLDOMReader.class.getName()).log(Level.SEVERE, null, ex);
+        }
+    }
+
+    protected void initDocument() {
+        try {
+            aDocument = aDocumentBuilder.parse(file);
+        } catch (SAXException ex) {
+            Logger.getLogger(Generic_XMLDOMReader.class.getName()).log(Level.SEVERE, null, ex);
+        } catch (IOException ex) {
+            Logger.getLogger(Generic_XMLDOMReader.class.getName()).log(Level.SEVERE, null, ex);
+        }
+        //optional, but recommended
+        //read this - http://stackoverflow.com/questions/13786607/normalization-in-dom-parsing-with-java-how-does-it-work
+        aDocument.getDocumentElement().normalize();
+        System.out.println("Root element :" + aDocument.getDocumentElement().getNodeName());
+    }
+
+    protected void initNodeList(String nodeName) {
+        nodeList = aDocument.getElementsByTagName(nodeName);
+    }
+
+    protected void readNodeListElements() {
+        int depth = 0;
+        for (int i = 0; i < nodeList.getLength(); i++) {
+            Node aNode = nodeList.item(i);
+            System.out.println("depth " + depth + ", node " + i);
+            System.out.println(aNode.getNodeName());
+            short aNodeType = aNode.getNodeType();
+            if (aNodeType == Node.ELEMENT_NODE) {
+                Element eElement = (Element) aNode;
+                NodeList childnodes = eElement.getChildNodes();
+                if (childnodes.getLength() == 0) {
+                    String attributeName = aNode.getNodeValue();
+                    String attributeValue = eElement.getElementsByTagName(attributeName).item(0).getTextContent();
+                    System.out.println("depth " + depth + ", node " + i);
+                    System.out.println("" + attributeName + " : " + attributeValue);
+                } else {
+                    readNodeListElements(childnodes, depth + 1);
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 92:103, 118:129.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.